File: 977.go
func sortedSquares(nums []int) []int {
	idx, neg, pos := 0, 0, 0
	result := make([]int, len(nums))

	for pos < len(nums) && nums[pos] < 0 {
		pos++
	}
	neg = pos - 1

	for idx < len(nums) {
		if neg >= 0 && (pos >= len(nums) || -nums[neg] < nums[pos]) {
			result[idx] = nums[neg] * nums[neg]
			neg--
		} else {
			result[idx] = nums[pos] * nums[pos]
			pos++
		}
		idx++
	}

	return result
}

File: 977.rs
impl Solution {
    pub fn sorted_squares(nums: Vec<i32>) -> Vec<i32> {
        let mut pos = 0;
        let mut neg = 0;
        
        let mut result = Vec::with_capacity(nums.len());

        for num in nums.iter() {
            if *num < 0 {
                pos += 1;
            } else {
                break;
            }
        }

        neg = pos - 1;
        
        while result.len() < nums.len() {
            if neg < nums.len() && (pos >= nums.len() || -nums[neg] < nums[pos]) {
                result.push(nums[neg] * nums[neg]);
                neg -= 1;
            } else {
                result.push(nums[pos] * nums[pos]);
                pos += 1;
            }
        }

        result
    }
}
